### PR TITLE
Update self-supervised losses implementation and tests

### PR DESCRIPTION
**Pull request by surajyadav-research**
## Summary of Improvements #1528 

This PR corrects several issues in the existing self-supervised losses (BYOL, SimSiam, DINO, Barlow Twins) and brings the implementations in line with the original papers. The previous versions were minimal but incomplete, and in several cases unsafe (silent broadcasting, missing stop-gradients, incorrect view handling, etc.).  
The new implementations add correct two-view behavior, strict validation, and safer numerical handling.

---

## BYOL

- **Two-view formulation**
  - *Before:* Only supported a single `(q, z)` direction; users had to manually build the symmetric two-view BYOL loss outside the function.
  - *Now:* Adds **single-direction and symmetric two-view** BYOL via `symmetric` and four projections, matching the paper.

- **Teacher gradients**
  - *Before:* Target projections were differentiable; forgetting to apply `stop_gradient` silently broke BYOL.
  - *Now:* Always **applies `lax.stop_gradient`** on target projections inside the loss.

- **Shape validation**
  - *Before:* No shape checks → silent broadcasting or mismatched tensors.
  - *Now:* **Validates shapes for all projection pairs**, failing fast on incorrect usage.

- **Cosine similarity & eps handling**
  - *Before:* Reimplemented cosine similarity manually and used hard-coded `finfo.eps`.
  - *Now:* Uses shared **`_regression.cosine_similarity`** and a dtype-safe, configurable `eps`.

---

## SimSiam

- **Symmetric loss support**
  - *Before:* Only computed `D(p1, z2)`; symmetric SimSiam required manual assembly.
  - *Now:* Supports **single-direction and symmetric two-view** SimSiam in one API.

- **Stop-gradient enforcement**
  - *Before:* Relied on the caller to apply `stop_gradient(z)`; forgetting it invalidates SimSiam.
  - *Now:* The loss **internally stops gradients** through the target projections.

- **Shape checks**
  - *Before:* No shape checks between predictor and target tensors.
  - *Now:* **Enforces matching shapes for all predictor/target pairs**.

- **Shared cosine similarity**
  - *Before:* Manual inline cosine implementation.
  - *Now:* Uses `_regression.cosine_similarity` with dtype-aligned `eps`.

---

## DINO

- **Single-view and two-view support**
  - *Before:* Only supported a single `(student, teacher)` pair; cross-view teacher/student matching had to be implemented manually.
  - *Now:* Adds **single-view and symmetric two-view DINO** via `two_view` and a second logit pair.

- **Teacher stop-gradient**
  - *Before:* Teacher probabilities were differentiable → gradients flowed into the EMA teacher.
  - *Now:* Wraps teacher softmax outputs in **`lax.stop_gradient`**.

- **Temperature and shape validation**
  - *Before:* Did not validate shapes or require temperatures to be positive.
  - *Now:* **Checks logits shapes** and **asserts positive temperatures** before use.

- **Centering & broadcasting**
  - *Before:* `center` subtraction relied on uncontrolled broadcasting.
  - *Now:* **Casts and broadcasts center explicitly**, ensuring consistent teacher normalization across views.

---

## Barlow Twins

- **Input shape & rank checks**
  - *Before:* Accepted arbitrary shapes; incorrect ranks led to meaningless correlation matrices.
  - *Now:* Enforces **matching shapes** and **rank-2 inputs** `[batch, feature_dim]`.

- **Numerically stable normalization**
  - *Before:* Used `jnp.std` with eps added after; less explicit numeric pathway.
  - *Now:* **Computes variance explicitly, adds eps before sqrt**, and normalizes safely.

- **Dtype-safe hyperparameters**
  - *Before:* `eps` and `lambda` were raw Python floats (problematic in mixed precision).
  - *Now:* **Casts both to the projection dtype** for consistent computation.

- **Clear loss decomposition**
  - *Before:* Diagonal and off-diagonal terms were computed but not clearly separated.
  - *Now:* Explicit **on-diag + off-diag** formulation directly matching the Barlow Twins paper.

---

## Tests

The test suite was updated to rigorously validate both the math and the new code paths.

### BYOL

- **Did not test symmetric path → Now tests exact symmetric formula with a handmade cosine reference.**
- **Only rough cosine check on random data → Now uses deterministic handmade comparison against the analytical expression.**
- **Weak JIT check (shape only) → Now verifies the JIT’ed loss matches the mathematical ground truth.**

### SimSiam

- **Never validated symmetric mode → Added an exact symmetric test that matches the cosine-based definition.**
- **Only tested trivial edge cases (identical/orthogonal) → Now tests the full objective with a precise reference implementation.**
- **No strict math-level comparison → Now enforces exact equivalence between implementation and the SimSiam loss definition.**

### DINO

- **No verification of the full KL / cross-entropy formula → Now compares against a handmade softmax/log-softmax reference.**
- **Two-view coupling never tested → Added explicit L12/L21 tests that compute both directions by hand.**
- **No argument validation tests → Added tests that temperatures must be positive and logit shapes must match.**

### Barlow Twins

- **Did not test the full normalization + cross-correlation pipeline → Now checks against a full handmade correlation computation.**
- **Previously only asserted non-negativity → Now validates the complete objective numerically, including on- and off-diagonal terms.**
- **No invalid-shape testing → Added tests for rank mismatches and shape mismatches to confirm error paths.**

---

### CODE REVIEW

**Comment by surajyadav-research** [comment]
Hi @rajasekharporeddy 
This PR is ready for review. I’ve updated the docs, implementation, and added/updated tests. All CI checks are green and CLA is done. Would you mind taking a look when you get a chance?

File: optax/losses/_self_supervised_test.py [hunk 1]
```diff
@@ -124,6 +124,362 @@ def test_vmap(self, anchor, positive, negative):
...
+    )
+    z2 = jnp.array(
+        [[0.0, -0.1, 0.4],
+         [0.2, 0.4, -0.3],
+         [0.9, -0.8, 0.3],
+         [-0.2, 0.6, 0.0]],
+        dtype=jnp.float32,
+    )
+
+    def testing_barlow_twins_loss(
```
Comment (rdyro) [inline optax/losses/_self_supervised_test.py]: instead of reimplementing the loss logic, can you compare e.g., that the loss is zero (and not nan) for identical inputs and non-zero otherwise?

File: optax/losses/_self_supervised_test.py [hunk 2]
```diff
@@ -124,6 +124,362 @@ def test_vmap(self, anchor, positive, negative):
...
+        student_logits,
+        teacher_logits,
+        student_temperature=student_temperature,
+        teacher_temperature=teacher_temperature,
+        teacher_center=teacher_center,
+        two_view=False,
+    )
+    np.testing.assert_allclose(result, handmade_result, atol=1e-6)
+
+  def test_two_view_matches_handmade(self):
```
Comment (rdyro) [inline optax/losses/_self_supervised_test.py]: rewrite as a random inputs test parameterized by a random seed please (instead of hard-coding the numbers)

File: optax/losses/_self_supervised_test.py [hunk 3]
```diff
@@ -124,6 +124,362 @@ def test_vmap(self, anchor, positive, negative):
...
+        p,
+        z,
+        symmetric=False,
+    )
+    np.testing.assert_allclose(result, handmade_result, atol=1e-4)
+
+
+class DinoLossTest(absltest.TestCase):
+
+  def test_single_view_matches_handmade(self):
```
Comment (rdyro) [inline optax/losses/_self_supervised_test.py]: rewrite as a random inputs test parameterized by a random seed please (instead of hard-coding the numbers)

File: optax/losses/_self_supervised_test.py [hunk 4]
```diff
@@ -124,6 +124,362 @@ def test_vmap(self, anchor, positive, negative):
...
+    result = jax.jit(_self_supervised.simsiam_loss)(
+        p1,
+        z2,
+        p2,
+        z1,
+        symmetric=True,
+    )
+    np.testing.assert_allclose(result, handmade_result, atol=1e-4)
+
+  def test_single_direction_matches_handmade(self):
```
Comment (rdyro) [inline optax/losses/_self_supervised_test.py]: rewrite as a random inputs test parameterized by a random seed please (instead of hard-coding the numbers)

File: optax/losses/_self_supervised_test.py [hunk 5]
```diff
@@ -124,6 +124,362 @@ def test_vmap(self, anchor, positive, negative):
...
+          epsilon=eps,
+      )
+      loss = -cos
+      return jnp.mean(loss)
+
+    handmade_result = testing_single_direction_simsiam(p, z)
+    result = jax.jit(_self_supervised.simsiam_loss)(
+        p,
+        z,
+        symmetric=False,
```
Comment (rdyro) [inline optax/losses/_self_supervised_test.py]: nit: too much whitespace, this can be on a single line

File: optax/losses/_self_supervised_test.py [hunk 6]
```diff
@@ -124,6 +124,362 @@ def test_vmap(self, anchor, positive, negative):
...
+        q,
+        z,
+        symmetric=False,
+    )
+    np.testing.assert_allclose(result, handmade_result, atol=1e-4)
+
+
+class SimSiamLossTest(absltest.TestCase):
+
+  def test_symmetric_batched_matches_handmade(self):
```
Comment (rdyro) [inline optax/losses/_self_supervised_test.py]: rewrite as a random inputs test parameterized by a random seed please (instead of hard-coding the numbers)

File: optax/losses/_self_supervised_test.py [hunk 7]
```diff
@@ -124,6 +124,362 @@ def test_vmap(self, anchor, positive, negative):
...
+    result = jax.jit(_self_supervised.byol_loss)(
+        q1,
+        z2,
+        q2,
+        z1,
+        symmetric=True,
+    )
+    np.testing.assert_allclose(result, handmade_result, atol=1e-4)
+
+  def test_single_direction_matches_handmade(self):
```
Comment (rdyro) [inline optax/losses/_self_supervised_test.py]: rewrite as a random inputs test parameterized by a random seed please (instead of hard-coding the numbers)

File: optax/losses/_self_supervised_test.py [hunk 8]
```diff
@@ -124,6 +124,362 @@ def test_vmap(self, anchor, positive, negative):
     np.testing.assert_allclose(vmap_loss.flatten(), original_loss.flatten()
                                , atol=1e-4)
 
+class ByolLossTest(absltest.TestCase):
+
+  def test_symmetric_batched_matches_handmade(self):
```
Comment (rdyro) [inline optax/losses/_self_supervised_test.py]: rewrite as a random inputs test parameterized by a random seed please (instead of hard-coding the numbers)

**Review by rdyro — COMMENTED** [review]
Thanks for the PR! I left some comments, mostly about improving the tests

**Comment by surajyadav-research** [comment]
Thanks for the review, @rdyro. I’ll make the requested updates ASAP.

**Comment by surajyadav-research** [comment]
Hi @rdyro 
I’ve updated the test code based on all your suggestions. However, I’m still seeing an import/copybara error and I’m not sure what’s causing it.

File: optax/losses/_self_supervised_test.py [hunk 9]
```diff
@@ -124,6 +124,239 @@ def test_vmap(self, anchor, positive, negative):
...
+  @parameterized.parameters(0, 1, 42)
+  def test_symmetric_batched_matches_handmade(self, seed):
+    key = jax.random.PRNGKey(seed)
+    k1, k2, k3, k4 = jax.random.split(key, 4)
+    q1 = jax.random.normal(k1, (2, 3), dtype=jnp.float32)
+    q2 = jax.random.normal(k2, (2, 3), dtype=jnp.float32)
+    z1 = jax.random.normal(k3, (2, 3), dtype=jnp.float32)
+    z2 = jax.random.normal(k4, (2, 3), dtype=jnp.float32)
+
+    def testing_byol_loss(q1_val, z2_val, q2_val, z1_val, eps=1e-6):
```
Comment (rdyro) [inline optax/losses/_self_supervised_test.py]: Let's ideally avoid a re-implementation tests, is there another way to test here?

File: optax/losses/_self_supervised_test.py [hunk 10]
```diff
@@ -124,6 +124,239 @@ def test_vmap(self, anchor, positive, negative):
...
+    np.testing.assert_allclose(result, handmade_result, atol=1e-4)
+
+  @parameterized.parameters(0, 1, 42)
+  def test_single_direction_matches_handmade(self, seed):
+    key = jax.random.PRNGKey(seed)
+    k1, k2 = jax.random.split(key, 2)
+    q = jax.random.normal(k1, (2, 3), dtype=jnp.float32)
+    z = jax.random.normal(k2, (2, 3), dtype=jnp.float32)
+
+    def testing_single_direction_byol(q_val, z_val, eps=1e-6):
```
Comment (rdyro) [inline optax/losses/_self_supervised_test.py]: Let's ideally avoid a re-implementation tests, is there another way to test here?

File: optax/losses/_self_supervised_test.py [hunk 11]
```diff
@@ -124,6 +124,239 @@ def test_vmap(self, anchor, positive, negative):
...
+  @parameterized.parameters(0, 1, 42)
+  def test_symmetric_batched_matches_handmade(self, seed):
+    key = jax.random.PRNGKey(seed)
+    k1, k2, k3, k4 = jax.random.split(key, 4)
+    p1 = jax.random.normal(k1, (2, 3), dtype=jnp.float32)
+    p2 = jax.random.normal(k2, (2, 3), dtype=jnp.float32)
+    z1 = jax.random.normal(k3, (2, 3), dtype=jnp.float32)
+    z2 = jax.random.normal(k4, (2, 3), dtype=jnp.float32)
+
+    def testing_simsiam_loss(p1_val, z2_val, p2_val, z1_val, eps=1e-6):
```
Comment (rdyro) [inline optax/losses/_self_supervised_test.py]: Let's ideally avoid a re-implementation tests, is there another way to test here?

File: optax/losses/_self_supervised_test.py [hunk 12]
```diff
@@ -124,6 +124,239 @@ def test_vmap(self, anchor, positive, negative):
...
+    np.testing.assert_allclose(result, handmade_result, atol=1e-4)
+
+  @parameterized.parameters(0, 1, 42)
+  def test_single_direction_matches_handmade(self, seed):
+    key = jax.random.PRNGKey(seed)
+    k1, k2 = jax.random.split(key, 2)
+    p = jax.random.normal(k1, (2, 3), dtype=jnp.float32)
+    z = jax.random.normal(k2, (2, 3), dtype=jnp.float32)
+
+    def testing_single_direction_simsiam(p_val, z_val, eps=1e-6):
```
Comment (rdyro) [inline optax/losses/_self_supervised_test.py]: Let's ideally avoid a re-implementation tests, is there another way to test here?

File: optax/losses/_self_supervised_test.py [hunk 13]
```diff
@@ -124,6 +124,239 @@ def test_vmap(self, anchor, positive, negative):
...
+  def test_single_view_matches_handmade(self, seed):
+    key = jax.random.PRNGKey(seed)
+    k1, k2, k3 = jax.random.split(key, 3)
+    student_logits = jax.random.normal(k1, (2, 3), dtype=jnp.float32)
+    teacher_logits = jax.random.normal(k2, (2, 3), dtype=jnp.float32)
+    student_temperature = 0.1
+    teacher_temperature = 0.04
+    teacher_center = jax.random.normal(k3, (3,), dtype=jnp.float32)
+
+    def testing_dino_loss(
```
Comment (rdyro) [inline optax/losses/_self_supervised_test.py]: Let's ideally avoid a re-implementation tests, is there another way to test here?

File: mlc_config.json
Comment (rdyro) [inline mlc_config.json:1]: What's the purpose of this file?

File: optax/losses/_self_supervised_test.py [hunk 14]
```diff
@@ -124,6 +124,239 @@ def test_vmap(self, anchor, positive, negative):
...
+
+    handmade_result = testing_byol_loss(q1, z2, q2, z1)
+    result = jax.jit(_self_supervised.byol_loss)(
+        q1, z2, q2, z1, symmetric=True,
+    )
+    np.testing.assert_allclose(result, handmade_result, atol=1e-4)
+
+  @parameterized.parameters(0, 1, 42)
+  def test_single_direction_matches_handmade(self, seed):
+    key = jax.random.PRNGKey(seed)
```
Comment (rdyro) [inline optax/losses/_self_supervised_test.py]: Prefer the modern `jax.random.key`

**Comment by surajyadav-research** [comment]
Hi @rdyro,
Ruff is flagging changes I didn’t make, coming from commit #1497  (pytype suppression comments on the lines above).
How should I handle this?

**Comment by rdyro** [comment]
PR [https://github.com/google-deepmind/optax/pull/1531](https://github.com/google-deepmind/optax/pull/1531) should be able to fix it once it's merged, can you rebase then?

Were you able to give it some more thought if there was a better way to test these losses instead of their re-implementation in the test file?

**Comment by surajyadav-research** [comment]
Hi @rdyro 
Yes, I’ve removed the redundant code in the tests and also implemented a few additional functions for better test coverage. I am confused due to the error I am encountering. I’ll push the new changes right away. These are changes i made.

### BYOL
- Added a single test that validates **symmetric (two-view)** behavior by comparing against the **average of two single-direction** calls.
- Added **zero-loss** sanity check when `q == z`.
- Added **scale invariance** check: `loss(a*q, b*z) == loss(q, z)` for any positive scalars `a, b`.
- Added **batch permutation invariance** check: `loss(q[perm], z[perm]) == loss(q, z)` for the same permutation `perm`.

### SimSiam
- Added a single test that validates **symmetric (two-view)** behavior by comparing against the **average of two single-direction** calls.
- Added **identity** sanity check: when `p == z`, loss is approximately `-1` (cosine similarity = 1).
- Added **scale invariance** check: `loss(a*p, b*z) == loss(p, z)` for any positive scalars `a, b`.
- Added **batch permutation invariance** check: `loss(p[perm], z[perm]) == loss(p, z)` for the same permutation `perm`.

### DINO
- Added **two-view** consistency test: `two_view=True` equals the average of two `two_view=False` calls with swapped pairs.
- Added sanity check that **student matches teacher** yields a lower/equal loss than a deterministic mismatch.
- Added **logit translation invariance** check: adding a constant offset to logits does not change the loss.

### Barlow Twins
- Added **batch permutation invariance** check: shuffling examples does not change the loss.

**Comment by rdyro** [comment]
Great, these sound like very good ideas for tests!

To unblock your flow, you could try rebasing your commits to the commit from Decemeber 26th [6b364ac289da913454bc849756f827f675de63c2](https://github.com/google-deepmind/optax/commit/6b364ac289da913454bc849756f827f675de63c2)

**Comment by surajyadav-research** [comment]
Hi @rdyro , I’ve pushed the updated test code. I tried to rebase it, but I’m getting stuck and couldn’t figure out the right steps. Sorry for the inconvenience.

**Comment by rdyro** [comment]
No worries, let's pick this up once the CI is unblocked.

**Comment by rdyro** [comment]
The CI fix is in, so you should be able to rebase to main now @surajyadav-research

**Comment by surajyadav-research** [comment]
Hi @rdyro 
I was checking the line that was causing the error, and it’s still present on `main`. When I rebase, that same line shows up in my local code as well.

**Comment by surajyadav-research** [comment]
Hi @rdyro ,
The code is now passing tests, and I’ve added the new tests I explained earlier.
The `mlc_config.json` file was causing timeouts during the tests, which is why it has been modified. I will squash all the commits into one shortly.
Whenever you have some time, could you please review the changes.
Thank you very much for your patience with me.

**Comment by surajyadav-research** [comment]
Hi @rdyro 
In this implementation, I’ve updated the tests I mentioned earlier.
When you get a chance, could you please review the changes?
